src/lazy/text/buffer.rs

```diff
@@ -1156,41 +1218,76 @@ impl<'data> TextBufferView<'data> {
+        self,
+        delimiter: u8,
+        allow_unescaped_newlines: bool,
+    ) -> IonParseResult<'data, (Self, bool)> {
         let mut contains_escaped_chars = false;
-        for (index, byte) in self.bytes().iter().enumerate() {
-            if is_escaped {
-                // If we're escaped, the previous byte was a \ and we ignore this one.
-                is_escaped = false;
-                continue;
-            }
-            if *byte == b'\\' {
-                is_escaped = true;
-                contains_escaped_chars = true;
-                continue;
-            }
-            if *byte == b'\r' {
-                // If the text contains an unescaped carriage return, we may need to normalize it.
-                // In some narrow cases, setting this flag to true may result in a sanitization buffer
-                // being allocated when it isn't strictly necessary.
+        // This de-sugared syntax allows us to modify `iter` mid-loop.
+        let mut iter = self.bytes().iter().copied().enumerate();
+        while let Some((index, byte)) = iter.next() {
+            if byte == b'\\' {
+                // It's an escape sequence. For the purposes of finding the end delimiter, we can
+                // skip the next 1 byte unless this is \r\n, in which case we need to skip two.
+                // Other escape sequences that are followed by more than one byte (e.g. \u and \U)
+                // are always followed by ASCII letters, which aren't used as delimiters.
                 contains_escaped_chars = true;
+                // Peek at the next two bytes to see if this is a \r\n
+                let next_two_bytes = self.bytes().get(index + 1..index + 3);
+                let bytes_to_skip = if next_two_bytes == Some(&[b'\r', b'\n']) {
+                    2
+                } else {
+                    1
+                };
+                // Eagerly skip the next iterator values
+                let _ = iter.nth(bytes_to_skip - 1);
                 continue;
             }
-            if *byte == delimiter {
+            if byte == delimiter {
                 let matched = self.slice(0, index);
                 let remaining = self.slice_to_end(index);
                 return Ok((remaining, (matched, contains_escaped_chars)));
             }
+            // If this is a control character, make sure it's a legal one.
+            if byte < 0x20 {
+                if byte == b'\r' {
+                    // Carriage returns are not actual escapes, but do require a substitution
+                    // as part of newline normalization when the string is read.
+                    contains_escaped_chars = true;
+                } else {
+                    self.validate_string_control_character(byte, index, allow_unescaped_newlines)?;
+                }
+            }
         }
         Err(nom::Err::Incomplete(Needed::Unknown))
     }
 
+    #[cold]
+    fn validate_string_control_character(
```

Comment:
🗺️ This is broken into its own `#[cold]` method because unescaped control characters in a string aren't that common, and the code to build up the errors adds up.